### PR TITLE
updates for php 7.2

### PR DESCRIPTION
**Pull request by qadan**
**JIRA Ticket**: https://jira.duraspace.org/browse/ISLANDORA-2414

# What does this Pull Request do?

A couple of updates for PHP 7.2+ compatibility

# What's new?

Smaller updates:

* A few of the functions using `each()` in `modules/citeproc/includes/convert.inc` have been updated to use PHP 7.2+ friendly functionality.
* `islandora_bibliography`'s untracked dependency on `citation_exporter` is now tracked

Larger update:

* TCPDF has been removed from this repo and re-implemented as a proper library. The existing version of TCPDF bundled with `citation_exporter` is now 7 years old and no longer PHP 7.2+ compatible. More recent versions of the repo claim support for PHP 7.3. Functionality appears to be the same with version 6.2.25

# How should this be tested?

Setup:
* Use PHP 7.2+ for testing
* Clone https://github.com/tecnickcom/TCPDF to your site's libraries folder.

Upon ingesting one or more citation objects:
* When using the citeproc module to render metadata, no difference should be encountered with metadata display before and after the update
* After creating a bookmark containing these citation objects, no difference should be encountered when exporting bookmarks as a PDF before and after the update.

# Additional Notes:

This update does require people to clone down the TCPDF library, so there is some tiny update overhead involved with this beyond just the standard pull/clear cache. This is pretty standard stuff though

# Interested parties
@DonRichards and @bryjbrown as the maintainers of the respective modules, also @Islandora/7-x-1-x-committers in general as this would require some intervention from systems folk upon updating

### CODE REVIEW

**Review by bondjimbond — DISMISSED** [review]
Seems to work just fine! But we will need an update hook to notify people that they have to install the new library.

**Comment by DiegoPino** [comment]
If we  are targeting a PHP 7.2 audience we need to make sure the stack works with PHP 7.2. We only test for 7.1 https://github.com/Islandora/islandora_scholar/blob/7.x/.travis.yml#L38

**Comment by qadan** [comment]
@DiegoPino was actually wondering about this; I don't know that we currently target 7.2 in any of our other `.travis.yaml` files; is this something we want to add?

@bryjbrown I'll get that update hook in 👍

**Comment by DiegoPino** [comment]
@qadan i think that is something we want to add if we have @Islandora/7-x-1-x-committers commitment to fix every issue that appears once we do that and also, we are willing to sacrifice old versions that simply contradict new needs. Just my opinion, committers involvement has been pretty slow (includes me too of course)

**Comment by bondjimbond** [comment]
> committers involvement has been pretty slow

I agree with @DiegoPino here... formally adding a test for PHP 7.2 might open up a rather large can of worms. And there's already a lot of standard Drupal that throws errors with PHP 7.2... 

I am very much on board with this specific pull request, because it's (a) giving us an updated version of TCPDF, (b) separating it from the module into /sites/all/libraries, which would make it simpler to update if needed, and (c) updating old code to current practice that's more understandable (`foreach` vs `each`).

Those are all good things even without thinking about PHP 7.2+ compatibility. I'd be all for making sure all of Islandora passes PHP 7.2 tests -- but we really don't have enough humans putting time into Islandora 7.x anymore. [I have a hard time getting even simple-but-necessary pull requests reviewed after a month](https://github.com/Islandora/islandora_batch/pull/111); it's a stretch to think we'd get enough people to overhaul all our modules to fit newer coding standards.

**Comment by jonathangreen** [comment]
Eventually I think someone will want to run Islandora on Ubuntu 18.04, which I believe is running PHP 7.2. Especially those fleeing the EOL of Ubuntu 14.04. 

Might be good to break it into its own ticket through, and see if anyone is interested in tackling that.

**Comment by bondjimbond** [comment]
@qadan Looks like we've still got references to tcpdf in Travis that need to be cut. This pull should pass once they're removed.

> $ mv $TRAVIS_BUILD_DIR/modules/exporter/lib/tcpdf $HOME
mv: cannot stat ‘/home/travis/build/Islandora/islandora_scholar/modules/exporter/lib/tcpdf’: No such file or directory

> 0.00s$ mv $HOME/tcpdf $TRAVIS_BUILD_DIR/modules/exporter/lib/
mv: cannot stat ‘/home/travis/tcpdf’: No such file or directory
The command "mv $HOME/tcpdf $TRAVIS_BUILD_DIR/modules/exporter/lib/" exited with 1.

**Comment by bondjimbond** [comment]
@qadan still failing..

```
$ mv $HOME/tcpdf $TRAVIS_BUILD_DIR/modules/exporter/lib/
mv: cannot stat ‘/home/travis/tcpdf’: No such file or directory
The command "mv $HOME/tcpdf $TRAVIS_BUILD_DIR/modules/exporter/lib/" exited with 1.
```

**Comment by qadan** [comment]
I'm silly, forgot the other bit

**Comment by DonRichards** [comment]
Wow, thanks for all of this!

**Comment by bondjimbond** [comment]
@qadan Still waiting on that update hook, I believe?

**Comment by qadan** [comment]
Completely missed that requirement. Update hook has been added 👍

**Review by jonathangreen — APPROVED** [review]
Looks good to me 👍🏻 
Starting merge clock ⏰

**Comment by jonathangreen** [comment]
This library update should probably end up in the release notes @whikloj, whats the best way to note that so it actually ends up in the release notes?

**Comment by DiegoPino** [comment]
@whikloj @jonathangreen this needs an update to this repo's README.md 
"imperative that the TCPDF library is installed in the site libraries folder in order for PDF citation exports to continue to properly"
and i would suggest a documentation JIRA ticket labeled/tagged for this release so updating the release notes won't escape our radar.

**Comment by jonathangreen** [comment]
@DiegoPino just so I understand, currently in place: 
  - Update hook to notify users during update
  - Readme requirements updated
  - Release notes for the release

You would additionally like to see this library called out in the readme as needing to be installed specifically separate from the requirements section?

**Comment by jonathangreen** [comment]
@bondjimbond I know your on vacation, but if you are seeing this can you dismiss your review on this (if you are happy with update) since the update hook has been added? Can't merge until your review is dismissed.

**Comment by manez** [comment]
If @bondjimbond doesn't respond by tomorrow and everyone else on the thread is cool with it, I can dismiss his review. Brandon's been a release manager and fought to get things in with a code freeze looming, so I'm pretty sure he'd be cool with us taking some steps to get this one done.

**Comment by jonathangreen** [comment]
@qadan are you able to address @DiegoPino comment about some additional docs in readme? I'd like to get this merged before the code freeze.

**Comment by qadan** [comment]
@qadan @DiegoPino Looks like there was already an update notices section in the README? I've thrown a thing in there to make it clear. I'll get that ticket tagged so it can be scooped up in the documentation pass.

File: README.md
Comment (DiegoPino) [inline README.md]: @qadan not to be picky, but wording could need an "and" there?
Maybe "...submodule previously included the TCPDF library **and** has been removed"

Comment (qadan) [reply README.md]: Giving this a language cleanup 👍

**Comment by jonathangreen** [comment]
@DiegoPino does the new language look okay to you? Want to get your approval before I start the merge clock on this one again.

**Comment by DiegoPino** [comment]
@jonathangreen i'm ok. Thanks @qadan !

**Comment by jonathangreen** [comment]
Awesome. Thanks @DiegoPino and @qadan. I'll try to get this one in tomorrow.

**Comment by manez** [comment]
@jonathangreen  shall I dismiss the review so you can go ahead with this one?

**Comment by jonathangreen** [comment]
Sure @manez, that would be great.

**Comment by manez** [comment]
It's done. Sorry @bondjimbond (but I hope the vacation is going great 😄 )

**Pull request merged**
